Class Cast Exception while accessing Map Fields in a DynamicMessage protobuf argument

Hello! Recently I've run into an interesting error and I think it might be a problem in how CEL-Java processes `DynamicMessage` protobuf types that use a map field. I've encountered this error in the enterprise code base that I normally work in, and have pulled out a "minimal" repro that should illustrate the point.

When I run a CEL expression which accesses a map field on a Dynamically parsed protobuf type, I seem to receive a class cast exception: 
```
java.lang.ClassCastException: class com.google.protobuf.DynamicMessage cannot be cast to class com.google.protobuf.MapEntry (com.google.protobuf.DynamicMessage and com.google.protobuf.MapEntry are in unnamed module of loader 'app')
```

Meanwhile, performing the same operation by parsing a byte string into generated code, and then feeding that into my CelExpression works fine (see "event2" in my repro code). 

 From inspecting the parsed object, it appears that using the generated code parser will compile the byte string into a MapField object during parsing, while the DynamicMessage parser will simply leave the MapEntry's as a list of DynamicMessages. It seems that this list of MapEntry's is perhaps what CEL is throwing the exception about, but that is speculation.

The protobuf I'm using for my simple repro is something like:
```proto
syntax = "proto3";

message EventA {
  bool property_bool = 5;
  map<string, string> map_str = 11;
}
```

While my code looks like:
```java
// imports...
public class Main {

    public static void main(String[] args) throws InvalidProtocolBufferException, ScriptException {
        Examples.EventA generatedEventA = Examples.EventA.getDefaultInstance().toBuilder().setPropertyBool(true).putMapStr("key", "value").putMapStr("key2", "value2").build();

        Descriptors.Descriptor descriptor = Examples.EventA.getDescriptor();
        DynamicMessage dynamicType = DynamicMessage.newBuilder(descriptor).build();

        // Referencing the map in the event created by the generated code parser, will go just fine.
        String workingExpression = "event.property_bool == true && event2.map_str['key'] == 'value'";

        // Referencing the map in the dynamically parsed event will fail with an exception.
        String failingExpression = "event.property_bool == true && event.map_str['key'] == 'value'";

        ScriptHost.ScriptBuilder scriptBuilder = ScriptHost.newBuilder().build().buildScript(failingExpression);
        scriptBuilder.withTypes(dynamicType.getDefaultInstanceForType());
        scriptBuilder.withDeclarations(
            Decls.newVar("event", Decls.newObjectType(dynamicType.getDescriptorForType().getFullName())),
            Decls.newVar("event2", Decls.newObjectType(Examples.EventA.getDescriptor().getFullName()))
            );
        Script script = scriptBuilder.build();

        DynamicMessage dynamicEventA = dynamicType.getParserForType().parseFrom(generatedEventA.toByteString());
        Examples.EventA parsedUsingGeneratedCodeEvent = Examples.EventA.parseFrom(generatedEventA.toByteString());

        // Here we pass the same event twice, once where it is parsed dynamically using the descriptor + DynamicMessage,
        // and again where we simply pass in the original event.
        Boolean result = script.execute(Boolean.class, Map.of("event", dynamicEventA, "event2", parsedUsingGeneratedCodeEvent));
        System.out.println(result);
    }
}
```

The full exception stack trace I get while running this code is: 
```
Exception in thread "main" org.projectnessie.cel.tools.ScriptExecutionException: java.lang.ClassCastException: class com.google.protobuf.DynamicMessage cannot be cast to class com.google.protobuf.MapEntry (com.google.protobuf.DynamicMessage and com.google.protobuf.MapEntry are in unnamed module of loader 'app')
	at org.projectnessie.cel.tools.Script.execute(Script.java:49)
	at com.company.Main.main(Main.java:32)
Caused by: java.lang.ClassCastException: class com.google.protobuf.DynamicMessage cannot be cast to class com.google.protobuf.MapEntry (com.google.protobuf.DynamicMessage and com.google.protobuf.MapEntry are in unnamed module of loader 'app')
	at org.projectnessie.cel.common.types.pb.FieldDescription.getValueFromField(FieldDescription.java:471)
	at org.projectnessie.cel.common.types.pb.FieldDescription.getField(FieldDescription.java:438)
	at org.projectnessie.cel.common.types.pb.PbObjectT.get(PbObjectT.java:77)
	at org.projectnessie.cel.interpreter.AttributeFactory.refResolve(AttributeFactory.java:1222)
	at org.projectnessie.cel.interpreter.AttributeFactory$StringQualifier.qualify(AttributeFactory.java:840)
	at org.projectnessie.cel.interpreter.AttributeFactory$AbsoluteAttribute.tryResolve(AttributeFactory.java:363)
	at org.projectnessie.cel.interpreter.AttributeFactory$MaybeAttribute.resolve(AttributeFactory.java:612)
	at org.projectnessie.cel.interpreter.Interpretable$EvalAttr.eval(Interpretable.java:1698)

	at org.projectnessie.cel.interpreter.Interpretable$EvalEq.eval(Interpretable.java:395)
	at org.projectnessie.cel.interpreter.Interpretable$EvalAnd.eval(Interpretable.java:343)
	at org.projectnessie.cel.Prog.eval(Prog.java:89)
	at org.projectnessie.cel.tools.Script.execute(Script.java:43)
	... 1 more
```

I've attached a zip of my IntelliJ project with the above example in case that is helpful.
[cel-repro.tar.gz](https://github.com/projectnessie/cel-java/files/8895279/cel-repro.tar.gz)

--- a/core/build.gradle.kts
+++ b/core/build.gradle.kts
@@ -14,10 +14,13 @@
  * limitations under the License.
  */
 
+import com.google.protobuf.gradle.protoc
+
 plugins {
   `java-library`
   signing
   id("com.diffplug.spotless")
+  id("com.google.protobuf")
   id("me.champeau.jmh")
   id("org.caffinitas.gradle.aggregatetestresults")
   id("org.caffinitas.gradle.testsummary")
@@ -48,6 +51,19 @@ dependencies {
 }
 
 jmh { jmhVersion.set(rootProject.extra["versionJmh"] as String) }
+
+sourceSets.test {
+  java.srcDir(project.buildDir.resolve("generated/source/proto/test/java"))
+  java.destinationDirectory.set(project.buildDir.resolve("classes/java/generatedTest"))
+}
+
+protobuf {
+  // Configure the protoc executable
+  protobuf.protoc {
+    // Download from repositories
+    artifact = "com.google.protobuf:protoc:${rootProject.extra["versionProtobuf"]}"
+  }
+}
 
 val testJar by
   configurations.creating {

--- a/core/src/main/java/org/projectnessie/cel/common/types/pb/FieldDescription.java
+++ b/core/src/main/java/org/projectnessie/cel/common/types/pb/FieldDescription.java
@@ -438,7 +438,6 @@ public final class FieldDescription extends Description {
     return getValueFromField(desc, (Message) target);
   }
 
-  @SuppressWarnings({"unchecked", "rawtypes"})
   public static Object getValueFromField(FieldDescriptor desc, Message message) {
 
     if (isWellKnownType(desc) && !message.hasField(desc)) {
@@ -466,10 +465,31 @@ public final class FieldDescription extends Description {
       //  There is no way to do a "message.getMapField(desc, key)" (aka a "reflective counterpart"
       //  for the generated map accessor methods like 'getXXXTypeOrThrow()'), too.
       if (v instanceof List) {
-        List<MapEntry> lst = (List) v;
-        Map map = new HashMap(lst.size() * 4 / 3 + 1);
-        for (MapEntry e : lst) {
-          map.put(e.getKey(), e.getValue());
+        List<?> lst = (List<?>) v;
+        Map<Object, Object> map = new HashMap<>(lst.size() * 4 / 3 + 1);
+        for (Object e : lst) {
+          Object key;
+          Object value;
+          if (e instanceof MapEntry) {
+            key = ((MapEntry<?, ?>) e).getKey();
+            value = ((MapEntry<?, ?>) e).getValue();
+          } else if (e instanceof DynamicMessage) {
+            DynamicMessage dynMsg = (DynamicMessage) e;
+            List<FieldDescriptor> fields = dynMsg.getDescriptorForType().getFields();
+            if (fields.size() == 2) {
+              key = dynMsg.getField(fields.get(0));
+              value = dynMsg.getField(fields.get(1));
+            } else {
+              throw new IllegalArgumentException(
+                  String.format(
+                      "Unexpected %s (%s) in list of map fields, dynamic message with != 2 fields",
+                      e.getClass(), e));
+            }
+          } else {
+            throw new IllegalArgumentException(
+                String.format("Unexpected %s (%s) in list of map fields", e.getClass(), e));
+          }
+          map.put(key, value);
         }
         v = map;
       }

--- a/core/src/test/java/org/projectnessie/cel/proto/ProtoTest.java
+++ b/core/src/test/java/org/projectnessie/cel/proto/ProtoTest.java
@@ -1,0 +1,97 @@
+/*
+ * Copyright (C) 2022 The Authors of CEL-Java
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.cel.proto;
+
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.projectnessie.cel.Env.newCustomEnv;
+import static org.projectnessie.cel.EnvOption.declarations;
+import static org.projectnessie.cel.EnvOption.types;
+import static org.projectnessie.cel.Library.StdLib;
+
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.DynamicMessage;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.projectnessie.cel.Ast;
+import org.projectnessie.cel.Env;
+import org.projectnessie.cel.Env.AstIssuesTuple;
+import org.projectnessie.cel.EnvOption;
+import org.projectnessie.cel.Program;
+import org.projectnessie.cel.Program.EvalResult;
+import org.projectnessie.cel.checker.Decls;
+import org.projectnessie.cel.common.types.pb.ProtoTypeRegistry;
+import org.projectnessie.cel.common.types.ref.Val;
+import org.projectnessie.cel.proto.tests.ProtoTestTypes;
+
+public class ProtoTest {
+  @ParameterizedTest
+  @ValueSource(
+      strings = {
+        "event.property_bool == true && event2.map_str['key'] == 'value'",
+        "event.property_bool == true && event.map_str['key'] == 'value'"
+      })
+  void protobufMaps(String expression) throws Exception {
+    ProtoTestTypes.EventA generatedEventA =
+        ProtoTestTypes.EventA.getDefaultInstance().toBuilder()
+            .setPropertyBool(true)
+            .putMapStr("key", "value")
+            .putMapStr("key2", "value2")
+            .build();
+
+    Descriptors.Descriptor descriptor = ProtoTestTypes.EventA.getDescriptor();
+    DynamicMessage dynamicType = DynamicMessage.newBuilder(descriptor).build();
+
+    List<EnvOption> envOptions = new ArrayList<>();
+    envOptions.add(StdLib());
+    envOptions.add(
+        declarations(
+            Decls.newVar(
+                "event", Decls.newObjectType(dynamicType.getDescriptorForType().getFullName())),
+            Decls.newVar(
+                "event2",
+                Decls.newObjectType(ProtoTestTypes.EventA.getDescriptor().getFullName()))));
+    envOptions.add(types(dynamicType.getDefaultInstanceForType()));
+
+    Env env = newCustomEnv(ProtoTypeRegistry.newRegistry(), envOptions);
+
+    AstIssuesTuple astIss = env.parse(expression);
+    assertThat(astIss).extracting(AstIssuesTuple::hasIssues).isEqualTo(FALSE);
+    Ast ast = astIss.getAst();
+
+    astIss = env.check(ast);
+    assertThat(astIss).extracting(AstIssuesTuple::hasIssues).isEqualTo(FALSE);
+
+    Program prg = env.program(ast);
+
+    DynamicMessage dynamicEventA =
+        dynamicType.getParserForType().parseFrom(generatedEventA.toByteString());
+    ProtoTestTypes.EventA parsedUsingGeneratedCodeEvent =
+        ProtoTestTypes.EventA.parseFrom(generatedEventA.toByteString());
+
+    Map<String, Object> arguments = new HashMap<>();
+    arguments.put("event", dynamicEventA);
+    arguments.put("event2", parsedUsingGeneratedCodeEvent);
+    EvalResult result = prg.eval(arguments);
+    Val val = result.getVal();
+    assertThat(val).extracting(Val::booleanValue).isEqualTo(TRUE);
+  }
+}

--- a/core/src/test/proto/proto-map.proto
+++ b/core/src/test/proto/proto-map.proto
@@ -1,0 +1,10 @@
+syntax = "proto3";
+
+option java_package = "org.projectnessie.cel.proto.tests";
+option java_outer_classname = "ProtoTestTypes";
+option java_generate_equals_and_hash = true;
+
+message EventA {
+  bool property_bool = 5;
+  map<string, string> map_str = 11;
+}

